mne/beamformer/_dics.py

```diff
@@ -1,32 +1,557 @@
+
+    if not return_generator:
+        stcs = list(stcs)
+
+    return stcs
+
+
+@verbose
+def apply_dics_csd(csd, filters, verbose=None):
+    """Apply Dynamic Imaging of Coherent Sources (DICS) beamformer weights.
+
+    Apply a previously computed DICS beamformer to a cross-spectral density
+    (CSD) object to estimate source power in time and frequency windows
+    specified in the CSD object [1]_.
+
+    Parameters
+    ----------
+    csd : instance of CrossSpectralDensity
+        The data cross-spectral density (CSD) matrices. A source estimate is
+        performed for each frequency or frequency-bin defined in the CSD
+        object.
+    filters : dict
+        DICS spatial filter (beamformer weights)
+        Filter weights returned from `make_dics`.
+    verbose : bool | str | int | None
+        If not ``None``, override default verbose level (see
+        :func:`mne.verbose` and :ref:`Logging documentation <tut_logging>` for
+        more).
+
+    Returns
+    -------
+    stc : SourceEstimate
+        Source power with frequency instead of time.
+    frequencies : list of float
+        The frequencies for which the source power has been computed. If the
+        data CSD object defines frequency-bins instead of exact frequencies,
+        the mean of each bin is returned.
+
+    References
+    ----------
+    .. [1] Gross et al. Dynamic imaging of coherent sources: Studying neural
+           interactions in the human brain. PNAS (2001) vol. 98 (2) pp. 694-699
+    """  # noqa: E501
+    ch_names = filters['ch_names']
+    vertices = filters['vertices']
+    n_orient = filters['n_orient']
+    subject = filters['subject']
+    n_sources = np.sum([len(v) for v in vertices])
+
+    # If CSD is summed over multiple frequencies, take the average frequency
+    frequencies = [np.mean(dfreq) for dfreq in csd.frequencies]
+    n_freqs = len(frequencies)
+
+    source_power = np.zeros((n_sources, len(csd.frequencies)))
+
+    # Ensure the CSD is in the same order as the weights
+    csd_picks = [csd.names.index(ch) for ch in ch_names]
+
+    logger.info('Computing DICS source power...')
+    for i, freq in enumerate(frequencies):
+        if n_freqs > 1:
+            logger.info('    applying DICS spatial filter at %sHz (%d/%d)' %
+                        (freq, i + 1, n_freqs))
+
+        Cm = csd.get_data(index=i)
+        Cm = Cm[csd_picks, :][:, csd_picks]
+        W = filters['weights'][i]
+
+        for k in range(n_sources):
+            Wk = W[n_orient * k: n_orient * k + n_orient]
+            power = Wk.dot(Cm).dot(Wk.T)
+
+            if n_orient > 1:  # Pool the orientations
+                source_power[k, i] = np.abs(power.trace())
+            else:
+                source_power[k, i] = np.abs(power)
+
+    logger.info('[done]')
+
+    return (_make_stc(source_power.reshape(-1, n_freqs), vertices=vertices,
+                      tmin=0, tstep=1, subject=subject), frequencies)
+
+
+@deprecated(deprecation_text)
```

Comment:
usually the `deprecated` does not go on private functions but public ones instead, so this should be moved to whatever public function(s) call(s) it